support/k8s/k8s-workload-registrar/controllers/spiffeid_controller.go

```diff
@@ -0,0 +1,295 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"path"
+
+	"github.com/sirupsen/logrus"
+	"github.com/spiffe/spire/pkg/common/idutil"
+	"github.com/spiffe/spire/proto/spire/api/registration"
+	"github.com/spiffe/spire/proto/spire/common"
+	"github.com/zeebo/errs"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	spiffeidv1beta1 "github.com/spiffe/spire/api/spiffecrd/v1beta1"
+)
+
+// SpiffeIDReconcilerConfig holds the config passed in when creating the reconciler
+type SpiffeIDReconcilerConfig struct {
+	Log          logrus.FieldLogger
+	R            registration.RegistrationClient
+	TrustDomain  string
+	Cluster      string
+	Mgr          ctrl.Manager
+}
+
+// SpiffeIDReconciler holds the runtime configuration and state of this controller
+type SpiffeIDReconciler struct {
+	client.Client
+	c                  SpiffeIDReconcilerConfig
+	myId               *string
+	spiffeIDCollection map[string]string
+}
+
+// NewSpiffeIDReconciler creates a new SpiffeIDReconciler object
+func NewSpiffeIDReconciler(config SpiffeIDReconcilerConfig) (*SpiffeIDReconciler, error) {
+	r := &SpiffeIDReconciler {
+		Client: config.Mgr.GetClient(),
+		c: config,
+		spiffeIDCollection:  make(map[string]string),
+	}
+
+	err := ctrl.NewControllerManagedBy(r.c.Mgr).
+		For(&spiffeidv1beta1.SpiffeID{}).
+		Complete(r)
+	if err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+// Initialize ensures there is a node registration entry for PSAT nodes in the cluster.
+func (r *SpiffeIDReconciler) Initialize(ctx context.Context) error {
+	nodeID := r.nodeID()
+	r.myId = &nodeID
+	return r.createEntry(ctx, &common.RegistrationEntry{
+		ParentId: idutil.ServerID(r.c.TrustDomain),
+		SpiffeId: nodeID,
+		Selectors: []*common.Selector{
+			{Type: "k8s_psat", Value: fmt.Sprintf("cluster:%s", r.c.Cluster)},
+		},
+	})
+}
+
+// Reconcile ensures the SPIRE Server entry matches the corresponding CRD
+func (r *SpiffeIDReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	spiffeID := spiffeidv1beta1.SpiffeID{}
+
+	if err := r.Get(ctx, req.NamespacedName, &spiffeID); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			r.c.Log.WithError(err).Error("unable to fetch SpiffeID")
+			return ctrl.Result{}, err
+		}
+
+		// Delete event
+		if err := r.ensureDeleted(ctx, r.spiffeIDCollection[req.NamespacedName.String()]); err != nil {
```

Comment:
Using a map to store the resource->entry mapping for deletes means that entries will be orphaned for any deletes that occur while the controller isn't running. Reintroducing the finaliser that was in the original controller guarantees that cleanup happens, and I think you can avoid  the pod deletion hanging if you set blockOwnerDeletion=false on the ownerReference for the pod.

Reply:
Good idea, ill try that out.

Reply:
this seems to mostly work. pod deletion wont hang with `blockOwnerDeletion` set to false. namespace deletion will still hang if there any deleted spiffeids in the namespace. I think its better than using in-memory maps regardless.

Reply:
committed changes to remove the spiffeIdCollection map and replaced with (mostly) non-blocking finalizer